### PR TITLE
Jaas secret builder

### PR DESCRIPTION
**Pull request by nandaloussiBabylon**
###  Add secret for jaas sasl config when a SCRAM user is created

_Select the type of your PR_
- Enhancement / new feature

### Description

this feature allows the user operator to not only create the SCRAM secret for a user but also a sasl jaas config string with the username and password

### CODE REVIEW

**Comment by scholzj** [comment]
Thanks for the PR. Given how easy it is to build it locally in the client, why is it really needed for the operator to generate it into the secret?

**Comment by nandaloussiBabylon** [comment]
We have a use case for our microservice environment where we're adopting a sidecar model to handle kafka trafic. Having  this secret for our scram credentials will make mounting that Config a lot easier / cleaner and will require no client side handling.

**Comment by scholzj** [comment]
The things I'm a bit concerned about:
* This is really useful only for Java. I do not think we want to end up maintaining some configurations for all the different clients available for Kafka.
* This is normally very easy to do on the client side as well. It is just concatenating several strings. What if one of the future Java or Kafka versions changes something on this? Than you will be stuck with whatever the operator provides. Doing it on the client seems much easier and much more flexible with regards to any changes in Java or in the client. 

That is why it seems to me that moving what is basically a single `String.format(...)` call from the client to the operator might not be the right move. It would be different if it was some complex encoding such as with the PEM certificates which is non-trivial to deal with. But here it is really just puting together a simple String.

File: user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java [hunk 1]
```diff
@@ -44,6 +44,7 @@
     private static final Logger log = LogManager.getLogger(KafkaUserModel.class.getName());
 
     public static final String KEY_PASSWORD = "password";
+    public static final String KEY_JAAS_CONFIG = "jaas_config";
```
Comment (scholzj) [inline user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java:47]: MAyve using `sasl.jaas.config` (the name of the related Kafka client property) would be better for this?

File: user-operator/src/test/java/io/strimzi/operator/user/model/KafkaUserModelTest.java [hunk 2]
```diff
@@ -233,8 +232,10 @@ public void testGenerateSecretGeneratesPasswordWhenNoUserSecretExists()    {
                         .withKubernetesManagedBy(KafkaUserModel.KAFKA_USER_OPERATOR_NAME)
                         .toMap()));
 
-        assertThat(generatedSecret.getData().keySet(), is(singleton(KafkaUserModel.KEY_PASSWORD)));
+        assertThat(generatedSecret.getData().keySet(), is(set(KafkaUserModel.KEY_PASSWORD, KafkaUserModel.KEY_JAAS_CONFIG)));
         assertThat(new String(Base64.getDecoder().decode(generatedSecret.getData().get(KafkaUserModel.KEY_PASSWORD))), is("aaaaaaaaaa"));
+        assertThat(new String(generatedSecret.getData().get(KafkaUserModel.KEY_JAAS_CONFIG)), is("org.apache.kafka.common.security.scram.ScramLoginModule required username=\"" + ResourceUtils.NAME + "\" password=\"YWFhYWFhYWFhYQ==\";"));
```
Comment (scholzj) [inline user-operator/src/test/java/io/strimzi/operator/user/model/KafkaUserModelTest.java:237]: I'm a bit confused by this ... why does it not need base64 decoding? And why is the password here not `the same as it is above?

File: user-operator/src/test/java/io/strimzi/operator/user/model/KafkaUserModelTest.java [hunk 3]
```diff
@@ -281,7 +283,7 @@ public void testGenerateSecretGeneratesPasswordFromExistingTlsSecret()    {
                         .withKubernetesManagedBy(KafkaUserModel.KAFKA_USER_OPERATOR_NAME)
                         .toMap()));
 
-        assertThat(generated.getData().keySet(), is(singleton("password")));
+        assertThat(generated.getData().keySet(), is(set("password", "jaas_config")));
```
Comment (scholzj) [inline user-operator/src/test/java/io/strimzi/operator/user/model/KafkaUserModelTest.java:286]: Can we add the test for the value here as well?

File: user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java (hunk 1 above)
Comment (samuel-hawker) [reply user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java:47]: I'm in agreement there. Better to closely mirror Kafka in this case.

File: user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java [hunk 4]
```diff
@@ -248,6 +251,18 @@ public void maybeGeneratePassword(PasswordGenerator generator, Secret userSecret
...
     }
 
+    /**
+     * Gerenate a SASL SSL Jaas Config String for SCRAM-SHA-512 secret
+     *
+     * @param username The KafkaPrincipal username
+     * @param password The SCRAM password for a given user.
+     * @return SASL jaas config string
+     */
+    public String saslJaasConfigBuilder(String username, String password) {
```
Comment (samuel-hawker) [inline user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java:261]: I'd argue this isn't really a builder
a method name of `saslJaasConfig` is probably sufficient, you pass in user and pass and out comes a SASL jaas config string

File: user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java [hunk 5]
```diff
@@ -248,6 +251,18 @@ public void maybeGeneratePassword(PasswordGenerator generator, Secret userSecret
 
     }
 
+    /**
+     * Gerenate a SASL SSL Jaas Config String for SCRAM-SHA-512 secret
+     *
+     * @param username The KafkaPrincipal username
+     * @param password The SCRAM password for a given user.
```
Comment (samuel-hawker) [inline user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java:258]: Can we put full stops at the end of each line of javadoc?

**Comment by scholzj** [comment]
There hasn't been much movement on this PR since more than one month. I will close it. If needed, it can be reopened later. I opened the issue #3332 to cover this functionality.

**Pull request closed without merging**